Remotion/Data/DomainObjects/Mapping/InlineTypeDefinitionVisitor.cs

```diff
@@ -0,0 +1,134 @@
+      public T? Visit (TypeDefinition typeDefinition)
+      {
+        ArgumentUtility.CheckNotNull("typeDefinition", typeDefinition);
+
+        return typeDefinition.Accept(this);
+      }
+
+      /// <inheritdoc />
+      public T? VisitClassDefinition (ClassDefinition classDefinition)
+      {
+        ArgumentUtility.CheckNotNull("classDefinition", classDefinition);
+
+        return _classDefinitionHandler(this, classDefinition);
+      }
+
+      public T? VisitInterfaceDefinition (InterfaceDefinition interfaceDefinition)
+      {
+        ArgumentUtility.CheckNotNull("interfaceDefinition", interfaceDefinition);
+
+        return _interfaceDefinitionHandler(this, interfaceDefinition);
+      }
+    }
+
+    private class TypeDefinitionVisitor : ITypeDefinitionVisitor
+    {
+      private readonly Action<ITypeDefinitionVisitor, ClassDefinition> _classDefinitionHandler;
+      private readonly Action<ITypeDefinitionVisitor, InterfaceDefinition> _interfaceDefinitionHandler;
+
+      public TypeDefinitionVisitor (
+          Action<ITypeDefinitionVisitor, ClassDefinition> classDefinitionHandler,
+          Action<ITypeDefinitionVisitor, InterfaceDefinition> interfaceDefinitionHandler)
+      {
+        ArgumentUtility.CheckNotNull(nameof(classDefinitionHandler), classDefinitionHandler);
+        ArgumentUtility.CheckNotNull(nameof(interfaceDefinitionHandler), interfaceDefinitionHandler);
+
+        _classDefinitionHandler = classDefinitionHandler;
+        _interfaceDefinitionHandler = interfaceDefinitionHandler;
+      }
+
+      /// <inheritdoc />
+      public void Visit (TypeDefinition typeDefinition)
+      {
+        ArgumentUtility.CheckNotNull("typeDefinition", typeDefinition);
+
+        typeDefinition.Accept(this);
+      }
+
+      /// <inheritdoc />
+      public void VisitClassDefinition (ClassDefinition classDefinition)
+      {
+        ArgumentUtility.CheckNotNull("classDefinition", classDefinition);
+
+        _classDefinitionHandler(this, classDefinition);
+      }
+
+      /// <inheritdoc />
+      public void VisitInterfaceDefinition (InterfaceDefinition interfaceDefinition)
+      {
+        ArgumentUtility.CheckNotNull("interfaceDefinition", interfaceDefinition);
+
+        _interfaceDefinitionHandler(this, interfaceDefinition);
+      }
+    }
+
+    public static T? Visit<T> (
+        TypeDefinition typeDefinition,
+        Func<ITypeDefinitionVisitor<T>, ClassDefinition, T> classDefinitionHandler,
+        Func<ITypeDefinitionVisitor<T>, InterfaceDefinition, T> interfaceDefinitionHandler)
+    {
+      ArgumentUtility.CheckNotNull(nameof(typeDefinition), typeDefinition);
+      ArgumentUtility.CheckNotNull(nameof(classDefinitionHandler), classDefinitionHandler);
+      ArgumentUtility.CheckNotNull(nameof(interfaceDefinitionHandler), interfaceDefinitionHandler);
+
+      var visitor = new TypeDefinitionVisitor<T>(classDefinitionHandler, interfaceDefinitionHandler);
+      return visitor.Visit(typeDefinition);
```

Comment:
The problem with not having `Visit (TypeDefinition)` on the visitor (see other comment) is solved by doing `typeDefinition.Accept(visitor)`